crates/ere-jolt/src/lib.rs

```diff
@@ -24,53 +28,73 @@ pub struct JOLT_TARGET;
-        let manifest_path = workspace_directory
-            .join(guest_relative)
-            .to_path_buf()
-            .join("Cargo.toml");
-        let package_name = package_name_from_manifest(&manifest_path).unwrap();
-        let mut program = Program::new(&package_name);
-        program.set_std(true);
-        program.set_manifest_path(manifest_path);
-
-        // TODO: Note that if this fails, it will panic which is why it doesn't return a Result.
-        program.build(DEFAULT_TARGET_DIR);
-
-        Ok(program)
+        let guest_dir = workspace_directory.join(guest_relative);
+
+        // Change current directory for `Program::build` to build guest program.
+        set_current_dir(&guest_dir).map_err(|source| CompileError::SetCurrentDirFailed {
+            source,
+            path: guest_dir.to_path_buf(),
+        })?;
+
+        let package_name = package_name_from_manifest(Path::new("Cargo.toml"))?;
+
+        // Note that if this fails, it will panic, hence we need to catch it.
+        let elf_path = std::panic::catch_unwind(|| {
+            let mut program = Program::new(&package_name);
+            program.set_std(true);
+            program.build(DEFAULT_TARGET_DIR);
+            program.elf.unwrap()
+        })
+        .map_err(|_| CompileError::BuildFailed)?;
+
+        Ok(elf_path)
     }
 }
 
+#[derive(CanonicalSerialize, CanonicalDeserialize)]
+pub struct EreJoltProof {
+    proof: JoltHyperKZGProof,
+    public_outputs: Vec<u8>,
+}
+
 pub struct EreJolt {
-    program: <JOLT_TARGET as Compiler>::Program,
+    program: jolt::host::Program,
+    prover_preprocessing: JoltProverPreprocessing<4, jolt::F, jolt::PCS, jolt::ProofTranscript>,
+    verifier_preprocessing: JoltVerifierPreprocessing<4, jolt::F, jolt::PCS, jolt::ProofTranscript>,
+    _resource: ProverResourceType,
 }
 
 impl EreJolt {
-    pub fn new(
-        program: <JOLT_TARGET as Compiler>::Program,
-        _resource_type: ProverResourceType,
-    ) -> Self {
-        EreJolt { program }
+    pub fn new(elf_path: PathBuf, _resource: ProverResourceType) -> Self {
+        let mut program = Program::new("");
+        program.elf = Some(elf_path);
+        let prover_preprocessing = preprocess_prover(&program);
+        let verifier_preprocessing = preprocess_verifier(&program);
+        EreJolt {
+            program,
+            prover_preprocessing,
+            verifier_preprocessing,
+            _resource,
+        }
     }
 }
 impl zkVM for EreJolt {
     fn execute(
         &self,
         _inputs: &Input,
     ) -> Result<zkvm_interface::ProgramExecutionReport, zkVMError> {
-        // TODO: check ProgramSummary
-        // TODO: FIXME
-        // let summary = self
-        //     .program
-        //     .clone()
-        //     .trace_analyze::<jolt::F>(inputs.bytes());
-        // let trace_len = summary.trace_len();
-        let trace_len = 0;
+        // TODO: Check how to pass private input to jolt
```

Comment:
Side question: Do we want to pass private inputs to zkVMs in ere? Prob we need to work a bit the `Input` type to reflect this?

Reply:
Sorry, I was thinking about zk. Yes, all inputs must be private.